// @rialo/frost-core/dist/index.js
import { createRialoClient } from '@rialo/ts-cdk';
export { RIALO_DEVNET_CHAIN, RIALO_LOCALNET_CHAIN, RIALO_MAINNET_CHAIN, RIALO_TESTNET_CHAIN, getDefaultRialoClientConfig } from '@rialo/ts-cdk';
import { Store } from '@tanstack/store';
import { getWallets as getWallets$1 } from '@wallet-standard/app';

// src/index.ts

// src/errors/base.ts
var FrostError = class extends Error {
  walletName;
  constructor(message, walletName, options) {
    super(message, options);
    this.name = this.constructor.name;
    this.walletName = walletName;
    const ErrorWithCapture = Error;
    if (ErrorWithCapture.captureStackTrace) {
      ErrorWithCapture.captureStackTrace(this, this.constructor);
    }
  }
};
function isFrostError(error) {
  return error instanceof FrostError;
}

// src/errors/errors.ts
var WalletNotFoundError = class extends FrostError {
  code = "WALLET_NOT_FOUND";
  constructor(walletName) {
    super(`Wallet "${walletName}" not found`, walletName);
  }
};
var WalletDisconnectedError = class extends FrostError {
  code = "WALLET_DISCONNECTED";
  constructor() {
    super("No wallet connected");
  }
};
var UnsupportedChainError = class extends FrostError {
  code = "UNSUPPORTED_CHAIN";
  chainId;
  supportedChains;
  constructor(chainId, supportedChains, walletName) {
    super(
      `Wallet "${walletName}" does not support chain "${chainId}". Supported: ${supportedChains.join(", ")}`,
      walletName
    );
    this.chainId = chainId;
    this.supportedChains = supportedChains;
  }
};
var UnsupportedFeatureError = class extends FrostError {
  code = "UNSUPPORTED_FEATURE";
  feature;
  constructor(feature, walletName) {
    super(`Wallet "${walletName}" does not support feature "${feature}"`, walletName);
    this.feature = feature;
  }
};
var ConnectionFailedError = class extends FrostError {
  code = "CONNECTION_FAILED";
  reason;
  constructor(reason, walletName, cause) {
    super(`Failed to connect to "${walletName}": ${reason}`, walletName, { cause });
    this.reason = reason;
  }
};
var SessionExpiredError = class extends FrostError {
  code = "SESSION_EXPIRED";
  expiredAt;
  constructor(expiredAt) {
    super("Session has expired");
    this.expiredAt = expiredAt;
  }
};
var WalletError = class extends FrostError {
  code = "WALLET_ERROR";
  originalError;
  constructor(originalError, walletName) {
    const message = originalError instanceof Error ? originalError.message : "Wallet operation failed";
    super(message, walletName, {
      cause: originalError instanceof Error ? originalError : void 0
    });
    this.originalError = originalError;
  }
};
var TransactionFailedError = class extends FrostError {
  code = "TRANSACTION_FAILED";
  signature;
  reason;
  constructor(signature, reason, walletName) {
    super(reason ?? "Transaction execution failed on-chain", walletName);
    this.signature = signature;
    this.reason = reason;
  }
};

// src/actions/connection.ts
function getActiveConnection(config) {
  const state = config.store.state;
  return {
    status: state.status,
    chainId: state.chainId,
    walletName: state.walletName,
    accountAddress: state.accountAddress,
    connectedAt: state.connectedAt,
    lastUsedAt: state.lastUsedAt
  };
}
var pendingConnections = /* @__PURE__ */ new WeakMap();
async function connect(config, options) {
  let walletConnections = pendingConnections.get(config);
  if (!walletConnections) {
    walletConnections = /* @__PURE__ */ new Map();
    pendingConnections.set(config, walletConnections);
  }
  const existingConnection = walletConnections.get(options.walletName);
  if (existingConnection) {
    return existingConnection;
  }
  const connectionPromise = connectInternal(config, options);
  walletConnections.set(options.walletName, connectionPromise);
  try {
    return await connectionPromise;
  } finally {
    walletConnections.delete(options.walletName);
  }
}
async function connectInternal(config, options) {
  const { walletName, silent = false } = options;
  const { store, options: configOpts } = config;
  const wallet = store.state._walletRefs.get(walletName);
  if (!wallet) {
    throw new WalletNotFoundError(walletName);
  }
  store.setState((state) => ({
    ...state,
    status: "connecting",
    walletName
  }));
  try {
    const connectFeature = wallet.features["standard:connect"];
    if (!connectFeature) {
      throw new UnsupportedFeatureError("standard:connect", walletName);
    }
    const result = await connectFeature.connect({ silent });
    if (!result.accounts.length) {
      throw new ConnectionFailedError("No accounts returned", walletName);
    }
    const account = result.accounts[0];
    const now = Date.now();
    store.setState((state) => {
      const accounts = new Map(state.accounts);
      accounts.set(account.address, {
        address: account.address,
        publicKey: new Uint8Array(account.publicKey),
        walletName,
        chains: [...account.chains],
        _raw: account
      });
      const wallets = new Map(state.wallets);
      const existingWallet = wallets.get(walletName);
      if (existingWallet) {
        wallets.set(walletName, {
          ...existingWallet,
          lastConnectedAt: now
        });
      }
      return {
        ...state,
        status: "connected",
        walletName,
        accountAddress: account.address,
        connectedAt: now,
        sessionExpiresAt: now + configOpts.sessionTTL,
        lastUsedAt: now,
        userDisconnected: false,
        // Clear the flag on successful connect
        accounts,
        wallets
      };
    });
    config._bridge?.subscribeToWallet(wallet);
    return {
      walletName,
      accountAddress: account.address
    };
  } catch (error) {
    store.setState((state) => ({
      ...state,
      status: "disconnected",
      walletName: null,
      accountAddress: null,
      connectedAt: null,
      sessionExpiresAt: null
    }));
    if (isFrostError(error)) {
      throw error;
    }
    throw new WalletError(error, walletName);
  }
}
async function disconnect(config) {
  const { store } = config;
  const { walletName } = store.state;
  if (!walletName) return;
  const wallet = store.state._walletRefs.get(walletName);
  config._bridge?.unsubscribeFromWallet(walletName);
  try {
    const disconnectFeature = wallet?.features["standard:disconnect"];
    if (disconnectFeature) {
      await disconnectFeature.disconnect();
    }
  } catch (error) {
    console.warn("[frost] Wallet disconnect failed:", error);
  }
  store.setState((state) => ({
    ...state,
    status: "disconnected",
    walletName: null,
    accountAddress: null,
    connectedAt: null,
    sessionExpiresAt: null,
    userDisconnected: true,
    accounts: /* @__PURE__ */ new Map()
  }));
}
function clearPendingConnections(config) {
  const walletConnections = pendingConnections.get(config);
  if (walletConnections) {
    walletConnections.clear();
    pendingConnections.delete(config);
  }
}
async function reconnect(config) {
  const { store } = config;
  const { walletName, sessionExpiresAt, userDisconnected } = store.state;
  if (!walletName) return null;
  if (userDisconnected) return null;
  if (sessionExpiresAt && sessionExpiresAt < Date.now()) {
    store.setState((state) => ({
      ...state,
      status: "disconnected",
      walletName: null,
      accountAddress: null,
      connectedAt: null,
      sessionExpiresAt: null
    }));
    return null;
  }
  store.setState((state) => ({ ...state, status: "reconnecting" }));
  try {
    const wallet = await config._registry?.waitForWallet(walletName, { timeout: 5e3 });
    if (!wallet) {
      throw new Error("Wallet not found");
    }
    return await connect(config, { walletName, silent: true });
  } catch {
    store.setState((state) => ({
      ...state,
      status: "disconnected",
      walletName: null,
      accountAddress: null,
      connectedAt: null,
      sessionExpiresAt: null
    }));
    return null;
  }
}

// src/storage/persistence.ts
var STORAGE_VERSION = 1;
function createPersistenceManager(key, storage) {
  let persistTimeout = null;
  return {
    persist(state) {
      if (persistTimeout) {
        clearTimeout(persistTimeout);
      }
      persistTimeout = setTimeout(() => {
        try {
          const serialized = {
            walletName: state.walletName,
            accountAddress: state.accountAddress,
            connectedAt: state.connectedAt,
            sessionExpiresAt: state.sessionExpiresAt,
            lastUsedAt: state.lastUsedAt,
            userDisconnected: state.userDisconnected,
            wallets: Array.from(state.wallets.entries()).map(([k, v]) => [
              k,
              {
                name: v.name,
                // icon intentionally excluded
                chains: [...v.chains],
                features: [...v.features],
                installedVersion: v.installedVersion,
                priority: v.priority,
                lastConnectedAt: v.lastConnectedAt
              }
            ]),
            accounts: Array.from(state.accounts.entries()).map(([k, v]) => [
              k,
              {
                address: v.address,
                publicKey: Array.from(v.publicKey),
                walletName: v.walletName,
                chains: [...v.chains],
                label: v.label
              }
            ])
          };
          const schema = {
            version: STORAGE_VERSION,
            state: serialized
          };
          storage.setItem(key, JSON.stringify(schema));
        } catch (error) {
          console.error("[Frost] Failed to persist state:", error);
        }
      }, 250);
    },
    clear() {
      storage.removeItem(key);
    },
    destroy() {
      if (persistTimeout) {
        clearTimeout(persistTimeout);
        persistTimeout = null;
      }
    }
  };
}
function loadPersistedState(key, storage, chainId) {
  try {
    const raw = storage.getItem(key);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    if (parsed.version !== STORAGE_VERSION) {
      storage.removeItem(key);
      return null;
    }
    const { state } = parsed;
    const now = Date.now();
    const sessionExpired = state.sessionExpiresAt && state.sessionExpiresAt < now;
    if (sessionExpired) {
      return {
        status: "disconnected",
        chainId,
        walletName: null,
        accountAddress: null,
        connectedAt: null,
        sessionExpiresAt: null,
        lastUsedAt: state.lastUsedAt,
        userDisconnected: false,
        wallets: new Map(
          state.wallets.map(([k, v]) => [
            k,
            {
              ...v,
              // icon will be populated when wallet is rediscovered
              chains: v.chains ?? [],
              features: v.features ?? []
            }
          ])
        ),
        accounts: /* @__PURE__ */ new Map()
        // Clear accounts on session expiry
      };
    }
    return {
      // Always load as disconnected - reconnect() will re-establish if appropriate
      status: "disconnected",
      // chainId comes from client, not persistence
      chainId,
      walletName: state.walletName,
      accountAddress: state.accountAddress,
      connectedAt: state.connectedAt,
      sessionExpiresAt: state.sessionExpiresAt,
      lastUsedAt: state.lastUsedAt,
      userDisconnected: state.userDisconnected ?? false,
      wallets: new Map(
        state.wallets.map(([k, v]) => [
          k,
          {
            ...v,
            // icon will be populated when wallet is rediscovered
            chains: v.chains ?? [],
            features: v.features ?? []
          }
        ])
      ),
      accounts: new Map(
        state.accounts.map(([k, v]) => [
          k,
          {
            ...v,
            publicKey: new Uint8Array(v.publicKey),
            chains: v.chains ?? []
            // Note: _raw is not persisted - will be populated on reconnect
          }
        ])
      )
    };
  } catch {
    return null;
  }
}

// src/config.ts
var DEFAULT_OPTIONS = {
  autoConnect: true,
  sessionTTL: 7 * 24 * 60 * 60 * 1e3,
  // 7 days
  storageKey: "rialo-frost",
  storage: typeof window !== "undefined" ? localStorage : null
};
function createConfig(options) {
  const opts = { ...DEFAULT_OPTIONS, ...options };
  let client = createRialoClient(opts.clientConfig);
  const chainId = opts.clientConfig.chain.id;
  const persistence = opts.storage ? createPersistenceManager(opts.storageKey, opts.storage) : null;
  const persisted = opts.storage ? loadPersistedState(opts.storageKey, opts.storage, chainId) : null;
  const initialState = persisted ? { ...persisted, _walletRefs: /* @__PURE__ */ new Map() } : {
    status: "disconnected",
    chainId,
    walletName: null,
    accountAddress: null,
    connectedAt: null,
    sessionExpiresAt: null,
    lastUsedAt: Date.now(),
    userDisconnected: false,
    wallets: /* @__PURE__ */ new Map(),
    accounts: /* @__PURE__ */ new Map(),
    _walletRefs: /* @__PURE__ */ new Map()
  };
  const store = new Store(initialState, {
    onUpdate: () => {
      persistence?.persist(store.state);
    }
  });
  const config = {
    store,
    get client() {
      return client;
    },
    getChainId() {
      return store.state.chainId;
    },
    switchChain(clientConfig) {
      client = createRialoClient(clientConfig);
      store.setState((state) => ({
        ...state,
        chainId: clientConfig.chain.id
      }));
    },
    _registry: null,
    // Will be initialized by initializeConfig
    _bridge: null,
    // Will be initialized by initializeConfig
    _persistence: persistence,
    options: opts,
    destroy() {
      clearPendingConnections(config);
      config._registry?.destroy();
      config._bridge?.destroy();
      config._persistence?.destroy();
    }
  };
  return config;
}
function initializeConfig(config, registry, bridge) {
  config._registry = registry;
  config._bridge = bridge;
}

// src/actions/accounts.ts
function getAccounts(config) {
  return Array.from(config.store.state.accounts.values());
}
function getActiveAccount(config) {
  const { accountAddress, accounts } = config.store.state;
  if (!accountAddress) return null;
  return accounts.get(accountAddress) ?? null;
}
function getAccount(config, address) {
  return config.store.state.accounts.get(address);
}

// src/actions/messages.ts
async function signMessage(config, options) {
  const { store } = config;
  const { walletName, accountAddress } = store.state;
  if (!walletName || !accountAddress) {
    throw new WalletDisconnectedError();
  }
  const wallet = store.state._walletRefs.get(walletName);
  if (!wallet) {
    throw new WalletNotFoundError(walletName);
  }
  const signFeature = wallet.features["rialo:signMessage"];
  if (!signFeature) {
    throw new UnsupportedFeatureError("rialo:signMessage", walletName);
  }
  const messageBytes = typeof options.message === "string" ? new TextEncoder().encode(options.message) : options.message;
  const account = store.state.accounts.get(accountAddress);
  if (!account?._raw) {
    throw new WalletDisconnectedError();
  }
  try {
    const results = await signFeature.signMessage({
      account: account._raw,
      message: messageBytes
    });
    store.setState((state) => ({
      ...state,
      lastUsedAt: Date.now()
    }));
    return {
      signature: results[0].signature,
      signedMessage: results[0].signedMessage
    };
  } catch (error) {
    if (isFrostError(error)) {
      throw error;
    }
    throw new WalletError(error, walletName);
  }
}

// src/actions/transactions.ts
async function signTransaction(config, options) {
  const { store } = config;
  const { walletName, accountAddress, chainId } = store.state;
  if (!walletName || !accountAddress) {
    throw new WalletDisconnectedError();
  }
  const wallet = store.state._walletRefs.get(walletName);
  if (!wallet) {
    throw new WalletNotFoundError(walletName);
  }
  const signFeature = wallet.features["rialo:signTransaction"];
  if (!signFeature) {
    throw new UnsupportedFeatureError("rialo:signTransaction", walletName);
  }
  const txBytes = "serialize" in options.transaction ? options.transaction.serialize() : options.transaction;
  const account = store.state.accounts.get(accountAddress);
  if (!account?._raw) {
    throw new WalletDisconnectedError();
  }
  try {
    const results = await signFeature.signTransaction({
      account: account._raw,
      transaction: txBytes,
      chain: chainId
    });
    store.setState((state) => ({
      ...state,
      lastUsedAt: Date.now()
    }));
    return {
      signedTransaction: results[0].signedTransaction
    };
  } catch (error) {
    if (isFrostError(error)) {
      throw error;
    }
    throw new WalletError(error, walletName);
  }
}
async function sendTransaction(config, options) {
  const { store } = config;
  const { walletName, accountAddress, chainId } = store.state;
  if (!walletName || !accountAddress) {
    throw new WalletDisconnectedError();
  }
  const wallet = store.state._walletRefs.get(walletName);
  if (!wallet) {
    throw new WalletNotFoundError(walletName);
  }
  const sendFeature = wallet.features["rialo:signAndSendTransaction"];
  if (!sendFeature) {
    throw new UnsupportedFeatureError("rialo:signAndSendTransaction", walletName);
  }
  const txBytes = "serialize" in options.transaction ? options.transaction.serialize() : options.transaction;
  const account = store.state.accounts.get(accountAddress);
  if (!account?._raw) {
    throw new WalletDisconnectedError();
  }
  try {
    const results = await sendFeature.signAndSendTransaction({
      account: account._raw,
      transaction: txBytes,
      chain: chainId
    });
    store.setState((state) => ({
      ...state,
      lastUsedAt: Date.now()
    }));
    return {
      signature: results[0].signature
    };
  } catch (error) {
    if (isFrostError(error)) {
      throw error;
    }
    throw new WalletError(error, walletName);
  }
}
async function signAndSendTransaction(config, options) {
  const { signedTransaction } = await signTransaction(config, {
    transaction: options.transaction
  });
  try {
    const result = await config.client.sendAndConfirmTransaction(
      signedTransaction,
      options.sendOptions
    );
    config.store.setState((state) => ({
      ...state,
      lastUsedAt: Date.now()
    }));
    if (!result.executed) {
      throw new TransactionFailedError(
        result.signature,
        result.err,
        config.store.state.walletName ?? void 0
      );
    }
    return { signature: result.signature };
  } catch (error) {
    if (isFrostError(error)) {
      throw error;
    }
    throw new WalletError(error, config.store.state.walletName ?? "unknown");
  }
}

// src/actions/wallets.ts
function getWallets(config) {
  return Array.from(config.store.state.wallets.values());
}
function getSortedWallets(config) {
  return Array.from(config.store.state.wallets.values()).sort((a, b) => {
    const priorityDiff = (b.priority ?? 0) - (a.priority ?? 0);
    if (priorityDiff !== 0) return priorityDiff;
    return (b.lastConnectedAt ?? 0) - (a.lastConnectedAt ?? 0);
  });
}
function getWallet(config, walletName) {
  return config.store.state.wallets.get(walletName);
}
function hasFeature(config, walletName, feature) {
  const wallet = config.store.state.wallets.get(walletName);
  return wallet?.features.includes(feature) ?? false;
}
function supportsChain(config, walletName) {
  const wallet = config.store.state.wallets.get(walletName);
  const chainId = config.store.state.chainId;
  return wallet?.chains.includes(chainId) ?? false;
}

// src/bridge/WalletEventBridge.ts
var WalletEventBridge = class {
  config;
  subscriptions = /* @__PURE__ */ new Map();
  constructor(config) {
    this.config = config;
  }
  subscribeToWallet(wallet) {
    const walletWithEvents = wallet;
    const eventsFeature = walletWithEvents.features["standard:events"];
    if (!eventsFeature) return;
    const unsubscribe = eventsFeature.on("change", (event) => {
      if (event.accounts) {
        this.handleAccountChange(wallet.name, event.accounts);
      }
    });
    this.subscriptions.set(wallet.name, unsubscribe);
  }
  unsubscribeFromWallet(walletName) {
    const unsubscribe = this.subscriptions.get(walletName);
    if (unsubscribe) {
      unsubscribe();
      this.subscriptions.delete(walletName);
    }
  }
  handleAccountChange(walletName, accounts) {
    const state = this.config.store.state;
    if (state.walletName !== walletName) return;
    if (accounts.length === 0) {
      this.config.store.setState((s) => ({
        ...s,
        status: "disconnected",
        walletName: null,
        accountAddress: null,
        connectedAt: null,
        sessionExpiresAt: null
      }));
      return;
    }
    const newAccount = accounts[0];
    this.config.store.setState((s) => {
      const accountsMap = new Map(s.accounts);
      accountsMap.set(newAccount.address, {
        address: newAccount.address,
        publicKey: new Uint8Array(newAccount.publicKey),
        walletName,
        chains: [...newAccount.chains],
        _raw: newAccount
      });
      return {
        ...s,
        accountAddress: newAccount.address,
        lastUsedAt: Date.now(),
        accounts: accountsMap
      };
    });
  }
  destroy() {
    for (const unsubscribe of this.subscriptions.values()) {
      unsubscribe();
    }
    this.subscriptions.clear();
  }
};
var RIALO_CHAIN_PREFIX = "rialo:";
var WalletRegistry = class {
  config;
  unsubscribe;
  readyPromise;
  readyResolve;
  constructor(config) {
    this.config = config;
    this.readyPromise = new Promise((resolve) => {
      this.readyResolve = resolve;
    });
    this.initialize();
  }
  initialize() {
    if (typeof window === "undefined") {
      this.readyResolve();
      return;
    }
    const { get, on } = getWallets$1();
    const existing = get();
    if (existing.length > 0) {
      let hasRialoWallet = false;
      for (const wallet of existing) {
        if (this.registerWallet(wallet)) {
          hasRialoWallet = true;
        }
      }
      if (hasRialoWallet) {
        this.readyResolve();
      }
    }
    const unsubscribeRegister = on("register", (wallet) => {
      if (this.registerWallet(wallet)) {
        this.readyResolve();
      }
    });
    const unsubscribeUnregister = on("unregister", (wallet) => {
      this.unregisterWallet(wallet.name);
    });
    this.unsubscribe = () => {
      unsubscribeRegister();
      unsubscribeUnregister();
    };
    setTimeout(() => this.readyResolve(), 1e3);
  }
  /**
   * Register a wallet if it supports Rialo chains
   * @returns true if wallet was registered (supports Rialo)
   */
  registerWallet(wallet) {
    if (!wallet.chains.some((c) => c.startsWith(RIALO_CHAIN_PREFIX))) {
      return false;
    }
    const entity = {
      name: wallet.name,
      icon: wallet.icon,
      chains: [...wallet.chains],
      features: Object.keys(wallet.features),
      installedVersion: wallet.version
    };
    this.config.store.setState((state) => {
      const wallets = new Map(state.wallets);
      const existing = wallets.get(wallet.name);
      wallets.set(wallet.name, {
        ...entity,
        lastConnectedAt: existing?.lastConnectedAt,
        priority: existing?.priority
      });
      const _walletRefs = new Map(state._walletRefs);
      _walletRefs.set(wallet.name, wallet);
      return { ...state, wallets, _walletRefs };
    });
    return true;
  }
  unregisterWallet(walletName) {
    this.config.store.setState((state) => {
      const wallets = new Map(state.wallets);
      wallets.delete(walletName);
      const _walletRefs = new Map(state._walletRefs);
      _walletRefs.delete(walletName);
      const shouldDisconnect = state.walletName === walletName;
      return {
        ...state,
        wallets,
        _walletRefs,
        ...shouldDisconnect && {
          status: "disconnected",
          walletName: null,
          accountAddress: null,
          connectedAt: null,
          sessionExpiresAt: null
        }
      };
    });
  }
  /** Wait for a specific wallet to be discovered */
  waitForWallet(name, options) {
    const timeout = options?.timeout ?? 5e3;
    const existing = this.config.store.state._walletRefs.get(name);
    if (existing) return Promise.resolve(existing);
    return new Promise((resolve) => {
      const timeoutId = setTimeout(() => {
        unsubscribe();
        resolve(null);
      }, timeout);
      const unsubscribe = this.config.store.subscribe(({ currentVal }) => {
        const wallet = currentVal._walletRefs.get(name);
        if (wallet) {
          clearTimeout(timeoutId);
          unsubscribe();
          resolve(wallet);
        }
      });
    });
  }
  /** Promise that resolves when at least one Rialo wallet is discovered */
  get ready() {
    return this.readyPromise;
  }
  /** Check if any wallets have been discovered */
  get isReady() {
    return this.config.store.state.wallets.size > 0;
  }
  destroy() {
    this.unsubscribe?.();
  }
};

export { ConnectionFailedError, FrostError, SessionExpiredError, TransactionFailedError, UnsupportedChainError, UnsupportedFeatureError, WalletDisconnectedError, WalletError, WalletEventBridge, WalletNotFoundError, WalletRegistry, connect, createConfig, disconnect, getAccount, getAccounts, getActiveAccount, getActiveConnection, getSortedWallets, getWallet, getWallets, hasFeature, initializeConfig, isFrostError, reconnect, sendTransaction, signAndSendTransaction, signMessage, signTransaction, supportsChain };
