// @rialo/frost-core/dist/testing.js
import { Store } from '@tanstack/store';

// src/testing/mocks.ts
var accountCounter = 0;
function createMockAccount(options = {}) {
  const id = ++accountCounter;
  const address = options.address ?? `MockAddress${id}${"X".repeat(32)}`.slice(0, 44);
  const publicKey = options.publicKey ?? new Uint8Array(32).fill(id);
  return {
    address,
    publicKey,
    chains: options.chains ?? ["rialo:devnet"],
    features: options.features ?? [],
    label: options.label,
    icon: void 0
  };
}
var walletCounter = 0;
function createMockWallet(options = {}) {
  const id = ++walletCounter;
  const {
    name = `MockWallet${id}`,
    icon = "data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg'/>",
    chains = ["rialo:devnet", "rialo:testnet"],
    accounts = [createMockAccount()],
    autoApprove = true,
    connectDelay = 0,
    failConnect = false,
    failDisconnect = false,
    supportsEvents = true,
    supportsSignMessage = true,
    supportsSignTransaction = true,
    supportsSignAndSend = true
  } = options;
  const eventListeners = /* @__PURE__ */ new Map();
  let connectCallCount = 0;
  let disconnectCallCount = 0;
  let signMessageCallCount = 0;
  let signTransactionCallCount = 0;
  let lastSignedMessage = null;
  let lastSignedTransaction = null;
  let connectedAccounts = [];
  const features = {};
  const connectFeature = {
    version: "1.0.0",
    connect: async ({ silent } = {}) => {
      connectCallCount++;
      if (connectDelay > 0) {
        await new Promise((resolve) => setTimeout(resolve, connectDelay));
      }
      if (failConnect) {
        throw failConnect instanceof Error ? failConnect : new Error("Connection rejected");
      }
      if (!autoApprove && !silent) {
        throw new Error("User rejected connection");
      }
      connectedAccounts = [...accounts];
      return { accounts: connectedAccounts };
    }
  };
  features["standard:connect"] = connectFeature;
  const disconnectFeature = {
    version: "1.0.0",
    disconnect: () => {
      disconnectCallCount++;
      if (failDisconnect) {
        throw failDisconnect instanceof Error ? failDisconnect : new Error("Disconnect failed");
      }
      connectedAccounts = [];
      return Promise.resolve();
    }
  };
  features["standard:disconnect"] = disconnectFeature;
  if (supportsEvents) {
    const eventsFeature = {
      version: "1.0.0",
      on: (event, listener) => {
        if (!eventListeners.has(event)) {
          eventListeners.set(event, /* @__PURE__ */ new Set());
        }
        eventListeners.get(event)?.add(listener);
        return () => {
          eventListeners.get(event)?.delete(listener);
        };
      }
    };
    features["standard:events"] = eventsFeature;
  }
  if (supportsSignMessage) {
    features["rialo:signMessage"] = {
      version: "1.0.0",
      signMessage: ({
        message
      }) => {
        signMessageCallCount++;
        lastSignedMessage = message;
        const signature = new Uint8Array(64);
        signature.fill(171);
        return Promise.resolve([
          {
            signature,
            signedMessage: message
          }
        ]);
      }
    };
  }
  if (supportsSignTransaction) {
    features["rialo:signTransaction"] = {
      version: "1.0.0",
      signTransaction: ({
        transaction
      }) => {
        signTransactionCallCount++;
        lastSignedTransaction = transaction;
        const signedTransaction = new Uint8Array(transaction.length + 64);
        signedTransaction.set(new Uint8Array(64).fill(205), 0);
        signedTransaction.set(transaction, 64);
        return Promise.resolve([{ signedTransaction }]);
      }
    };
  }
  if (supportsSignAndSend) {
    features["rialo:signAndSendTransaction"] = {
      version: "1.0.0",
      signAndSendTransaction: ({
        transaction
      }) => {
        signTransactionCallCount++;
        lastSignedTransaction = transaction;
        const signature = new Uint8Array(64);
        signature.fill(239);
        return Promise.resolve([{ signature }]);
      }
    };
  }
  const wallet = {
    version: "1.0.0",
    name,
    icon,
    chains,
    features,
    accounts: connectedAccounts,
    // Test helpers
    _emitAccountChange: (newAccounts) => {
      connectedAccounts = newAccounts;
      const listeners = eventListeners.get("change");
      if (listeners) {
        for (const listener of listeners) {
          listener({ accounts: newAccounts });
        }
      }
    },
    get _connectCallCount() {
      return connectCallCount;
    },
    get _disconnectCallCount() {
      return disconnectCallCount;
    },
    get _signMessageCallCount() {
      return signMessageCallCount;
    },
    get _signTransactionCallCount() {
      return signTransactionCallCount;
    },
    get _lastSignedMessage() {
      return lastSignedMessage;
    },
    get _lastSignedTransaction() {
      return lastSignedTransaction;
    },
    _reset: () => {
      connectCallCount = 0;
      disconnectCallCount = 0;
      signMessageCallCount = 0;
      signTransactionCallCount = 0;
      lastSignedMessage = null;
      lastSignedTransaction = null;
      connectedAccounts = [];
      eventListeners.clear();
    }
  };
  return wallet;
}
function createMockClient(options = {}) {
  const {
    defaultBalance = 1000000000n,
    // 1 RLO equivalent
    balances = /* @__PURE__ */ new Map(),
    failRequests = false,
    requestDelay = 0,
    transactionSignature = "MockTxSignature123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefgh"
  } = options;
  let sendTransactionCallCount = 0;
  let sendAndConfirmCallCount = 0;
  let lastSentTransaction = null;
  let transactionFails = false;
  let transactionError;
  const maybeDelay = async () => {
    if (requestDelay > 0) {
      await new Promise((resolve) => setTimeout(resolve, requestDelay));
    }
  };
  const maybeThrow = () => {
    if (failRequests) {
      throw failRequests instanceof Error ? failRequests : new Error("RPC request failed");
    }
  };
  const client = {
    getBalance: async (pubkey) => {
      await maybeDelay();
      maybeThrow();
      const address = pubkey.toString();
      return balances.get(address) ?? defaultBalance;
    },
    sendTransaction: async (transaction) => {
      await maybeDelay();
      maybeThrow();
      sendTransactionCallCount++;
      lastSentTransaction = transaction;
      return transactionSignature;
    },
    sendAndConfirmTransaction: async (transaction) => {
      await maybeDelay();
      maybeThrow();
      sendAndConfirmCallCount++;
      lastSentTransaction = transaction;
      return {
        signature: transactionSignature,
        executed: !transactionFails,
        err: transactionFails ? transactionError : void 0
      };
    },
    // Add other common RPC methods as needed
    getTransaction: async () => {
      await maybeDelay();
      maybeThrow();
      return { signature: transactionSignature, slot: 12345 };
    },
    getLatestBlockhash: async () => {
      await maybeDelay();
      maybeThrow();
      return {
        blockhash: "MockBlockhash123456789ABCDEFGHIJKLMNOPQRSTUV",
        lastValidBlockHeight: 1e5
      };
    },
    // Test helpers
    _setBalance: (address, balance) => {
      balances.set(address, balance);
    },
    get _sendTransactionCallCount() {
      return sendTransactionCallCount;
    },
    get _sendAndConfirmCallCount() {
      return sendAndConfirmCallCount;
    },
    get _lastSentTransaction() {
      return lastSentTransaction;
    },
    _setTransactionFails: (fails, err) => {
      transactionFails = fails;
      transactionError = err;
    },
    _reset: () => {
      sendTransactionCallCount = 0;
      sendAndConfirmCallCount = 0;
      lastSentTransaction = null;
      transactionFails = false;
      transactionError = void 0;
      balances.clear();
    }
  };
  return client;
}
function createMockStorage() {
  const store = /* @__PURE__ */ new Map();
  return {
    get length() {
      return store.size;
    },
    key(index) {
      const keys = Array.from(store.keys());
      return keys[index] ?? null;
    },
    getItem(key) {
      return store.get(key) ?? null;
    },
    setItem(key, value) {
      store.set(key, value);
    },
    removeItem(key) {
      store.delete(key);
    },
    clear() {
      store.clear();
    }
  };
}
function createMockConfig(options = {}) {
  const chainId = options.chainId ?? "rialo:devnet";
  const mockClient = createMockClient(options.clientOptions);
  const mockStorage = options.storage === null ? null : options.storage ?? createMockStorage();
  const initialState = {
    status: "disconnected",
    chainId,
    walletName: null,
    accountAddress: null,
    connectedAt: null,
    sessionExpiresAt: null,
    lastUsedAt: Date.now(),
    userDisconnected: false,
    wallets: /* @__PURE__ */ new Map(),
    accounts: /* @__PURE__ */ new Map(),
    _walletRefs: /* @__PURE__ */ new Map(),
    ...options.initialState
  };
  const store = new Store(initialState);
  const config = {
    store,
    // Cast to unknown first since MockRialoClient is a test-only partial implementation
    client: mockClient,
    getChainId: () => store.state.chainId,
    switchChain: () => {
    },
    _registry: null,
    _bridge: null,
    _persistence: null,
    options: {
      clientConfig: {},
      autoConnect: options.autoConnect ?? false,
      sessionTTL: 7 * 24 * 60 * 60 * 1e3,
      storageKey: "rialo-frost-test",
      storage: mockStorage
    },
    destroy: () => {
    }
  };
  return { config, mockClient, mockStorage };
}
async function waitFor(condition, options = {}) {
  const { timeout = 5e3, interval = 50 } = options;
  const startTime = Date.now();
  while (Date.now() - startTime < timeout) {
    if (await condition()) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, interval));
  }
  throw new Error(`waitFor timed out after ${timeout}ms`);
}
async function waitForState(config, predicate, options) {
  await waitFor(() => predicate(config.store.state), options);
  return config.store.state;
}
function createMockWalletEntity(overrides = {}) {
  return {
    name: "MockWallet",
    icon: "data:image/svg+xml,<svg/>",
    chains: ["rialo:devnet"],
    features: [
      "standard:connect",
      "standard:disconnect",
      "rialo:signMessage",
      "rialo:signTransaction"
    ],
    installedVersion: "1.0.0",
    priority: 0,
    lastConnectedAt: void 0,
    ...overrides
  };
}
function createMockAccountEntity(overrides = {}) {
  return {
    address: "MockAddress123456789ABCDEFGHIJKLMNOPQRSTUV",
    publicKey: new Uint8Array(32).fill(1),
    walletName: "MockWallet",
    chains: ["rialo:devnet"],
    label: void 0,
    _raw: void 0,
    ...overrides
  };
}

export { createMockAccount, createMockAccountEntity, createMockClient, createMockConfig, createMockStorage, createMockWallet, createMockWalletEntity, waitFor, waitForState };
